examples/ransomware_detection/stages/preprocessing.py

```diff
@@ -0,0 +1,198 @@
+        """
+        Create snapshot_id's sliding sequence for a given window
+        """
+
+        sliding_window_offsets = []
+
+        for start in range(ids_len - (window - 1)):
+            stop = start + window
+            sequence = ids[start:stop]
+            consecutive = sorted(sequence) == list(range(min(sequence), max(sequence) + 1))
+            if consecutive:
+                sliding_window_offsets.append((start, stop))
+
+        return sliding_window_offsets
+
+    def _rollover_pending_snapshots(self,
+                                    snapshot_ids: typing.List[int],
+                                    source_pid_process: str,
+                                    snapshot_df: pd.DataFrame):
+        """
+        Store the unprocessed snapshots from current run to a stateful member to process them in the next run.
+        """
+
+        pending_snapshots = []
+
+        for snapshot_id in snapshot_ids[1 - self._sliding_window:]:
+            pending_snapshot_data = snapshot_df[snapshot_df.index == snapshot_id].values[0]
+            pending_snapshot = SnapshotData(snapshot_id, pending_snapshot_data)
+            pending_snapshots.append(pending_snapshot)
+
+        if pending_snapshots:
+            self._snapshot_dict[source_pid_process] = pending_snapshots
+
+    def _merge_curr_and_prev_snapshots(self, snapshot_df: pd.DataFrame, source_pid_process: str) -> pd.DataFrame:
+        """
+        Merge current run snapshots with previous unprocessed snapshots.
+        """
+
+        prev_pending_snapshots = self._snapshot_dict[source_pid_process]
+
+        # If previous pending snapshots that exists in the memory. Just get them to process in this run.
+        for prev_pending_snapshot in prev_pending_snapshots:
+            snapshot_df.loc[prev_pending_snapshot.snapshot_id] = prev_pending_snapshot.data
+
+        # Keep snapshot_ids in order to generate sequence.
+        snapshot_df = snapshot_df.sort_index()
+
+        return snapshot_df
+
+    def _pre_process_batch(self, x: MultiMessage) -> MultiInferenceFILMessage:
+        """
+        This function is invoked for every source_pid_process.
+        It looks for any pending snapshots related to the source and pid process in the memory.
+        If there are any unprocessed snapshots in the memory, they are merged with existing snapshots,
+        and a series of snapshot features generatd based on the specified sliding window,
+        followed by the creation of an inference memory message.
+        Current run's unprocessed snapshots will be rolled over to the next.
+        """
+
+        snapshot_df = x.get_meta()
+
+        # print(snapshot_df[['snapshot_id','source_pid_process']])
+
+        curr_snapshots_size = len(snapshot_df)
+
+        # Set snapshot_id as index this is used to get ordered snapshots based on sliding window.
+        snapshot_df.index = snapshot_df.snapshot_id
```

Comment:
Is this updating `index` in-place on the MultiMessage, or is this creating a copy? If it's updating in place, don't we want to use `set_meta`?

Reply:
Here in this scenario we don't want to update entries in the MessageMeta to preserve generated features. So I have got the copy of MultiMessage entries and updating it to create a consecutive sequence using pending snapshots (If there are any) and current snapshots with respect to `source_pid_process`. Does it make sense?